src/net_processing.cpp

```diff
@@ -1782,30 +1880,63 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             }
         }
 
+        // When we succeed in decoding a block's txids from a cmpctblock
+        // message we typically jump to the BLOCKTXN handling code, with a
+        // dummy (empty) BLOCKTXN message, to re-use the logic there in
+        // completing processing of the putative block (without cs_main).
+        bool fProcessBLOCKTXN = false;
+        CDataStream blockTxnMsg(SER_NETWORK, PROTOCOL_VERSION);
+
+        // If we end up treating this as a plain headers message, call that as well
+        // without cs_main.
+        bool fRevertToHeaderProcessing = false;
+        CDataStream vHeadersMsg(SER_NETWORK, PROTOCOL_VERSION);
+
         // Keep a CBlock for "optimistic" compactblock reconstructions (see
         // below)
         std::shared_ptr<CBlock> pblock = std::make_shared<CBlock>();
         bool fBlockReconstructed = false;
 
+        {
         LOCK(cs_main);
         // If AcceptBlockHeader returned true, it set pindex
         assert(pindex);
         UpdateBlockAvailability(pfrom->GetId(), pindex->GetBlockHash());
 
-        std::map<uint256, pair<NodeId, list<QueuedBlock>::iterator> >::iterator blockInFlightIt = mapBlocksInFlight.find(pindex->GetBlockHash());
-        bool fAlreadyInFlight = blockInFlightIt != mapBlocksInFlight.end();
+        std::pair<BlockDownloadMap::iterator, BlockDownloadMap::iterator> rangeInFlight = mmapBlocksInFlight.equal_range(pindex->GetBlockHash());
+        bool fAlreadyInFlight = rangeInFlight.first != rangeInFlight.second;
+        bool fInFlightFromSamePeer = false;
+        bool fPeerWasFirstRequest = true;
+        bool peerFound = false;
+        size_t countPartialBlocksStarted = 0;
+        while (rangeInFlight.first != rangeInFlight.second) {
+            if (rangeInFlight.first->second.first == pfrom->GetId()) {
+                fInFlightFromSamePeer = true;
+                peerFound = true;
+            }
+            if (rangeInFlight.first->second.second->partialBlock)
+                countPartialBlocksStarted++;
+            rangeInFlight.first++;
+            if (!peerFound)
+                fPeerWasFirstRequest = false;
+        }
 
         if (pindex->nStatus & BLOCK_HAVE_DATA) // Nothing to do here
             return true;
-
+        LogPrintf("ChainWork check at height %d new: %s  tip: %s\n",pindex->nHeight,pindex->nChainWork.GetHex(),chainActive.Tip()->nChainWork.GetHex());
         if (pindex->nChainWork <= chainActive.Tip()->nChainWork || // We know something better
```

Comment:
why not make this < instead of <= ? why do we avoid requesting compact blocks for competing best blocks?